Tools/SetupUnityPackage/UnityUtils/FileHelper.cs

```diff
@@ -33,16 +35,35 @@ public static string GetInternalStorage()
 
             // On Android persistentDataPath returns the external folder, where the app may not have permissions to write.
             // we use reflection to call File.getAbsolutePath(currentActivity.getFilesDir())
-            using (var unityPlayer = new AndroidJavaClass("com.unity3d.player.UnityPlayer"))
-            using (var currentActivity = unityPlayer.GetStatic<AndroidJavaObject>("currentActivity"))
+            var path = string.Empty;
+
+            try
             {
-                var getFilesDir = AndroidJNIHelper.GetMethodID(AndroidJNI.FindClass("android/content/ContextWrapper"), "getFilesDir", "()Ljava/io/File;");
-                var internalFilesDir = AndroidJNI.CallObjectMethod(currentActivity.GetRawObject(), getFilesDir, Array.Empty<jvalue>());
-                var getAbsolutePath = AndroidJNIHelper.GetMethodID(AndroidJNI.FindClass("java/io/File"), "getAbsolutePath", "()Ljava/lang/String;");
+                if (AndroidJNI.AttachCurrentThread() == 0)
+                {
+                    using (var unityPlayer = new AndroidJavaClass("com.unity3d.player.UnityPlayer"))
+                    using (var currentActivity = unityPlayer.GetStatic<AndroidJavaObject>("currentActivity"))
+                    {
+                        var getFilesDir = AndroidJNIHelper.GetMethodID(AndroidJNI.FindClass("android/content/ContextWrapper"), "getFilesDir", "()Ljava/io/File;");
+                        var internalFilesDir = AndroidJNI.CallObjectMethod(currentActivity.GetRawObject(), getFilesDir, Array.Empty<jvalue>());
+                        var getAbsolutePath = AndroidJNIHelper.GetMethodID(AndroidJNI.FindClass("java/io/File"), "getAbsolutePath", "()Ljava/lang/String;");
```

Comment:
All these AndroidJNI methods return pointers. We could check for their "nullity" (IntPtr.Zero), but everything is a try and catch that does nothing if exceptions are raised. We can't do much if some of those methods return "null"/break, so I think it's safe to leave it as is.

Reply:
After reading up on it a bit, I don't think we need to use JNI here at all. These methods are all defined on the Java classes, so can't we do something like 

```
var filesDir = currentActivity.Call<AndroidJavaObject>("getFilesDir");
return filesDir.Call<string>("getAbsolutePath");
```